### DISCUSSION TITLE
STM324X1 blackpill experimental versatile GRBL-breakout board

### DISCUSSION THREAD

**Discussion by wolfgangr** [grblHAL/core#562] [Show and tell]
Please allow me to introduce my board design:

Design rationale:
- combine the flexibility of a breadboard setup with the clean design and distilled experience of a pre-made board
- just throw loads of jumper pins over the board, so anything might either be bridged as planned, left away or routed somewhere else by duPont lines
- blackpill STM32F411 at the heart:  large enough to serve many machine's needs ... and awful cheap

I started off with Terje Io's Nucleo board - simply because it was in KiCAD 5, which I'm still using.
https://github.com/terjeio/CNC_Breakout_Nucleo64

In the end, most of the hardware was closer to
https://github.com/phil-barrett/grblHAL-teensy-4.x
simply because that was what I had in mind.

Yesterday, the first batch arrived from the fab, but I haven't yet tested anything.

Features:
- Socket for a blackpill (don't see a reason to use bluepill any more)
- 10 optocoupler inputs for EMC robust limit switches and user buttons
- plus 2 pluggable reserve just since they were on the chips
- 2 schmitt trigger digital inputs intended for e.g. spindle encoder
- 2 analog inputs with proteciton circuits to map 10 V to 4000 AD-steps - or configurable for other needs
- a PWM-to10V lowpass driver (i think from phil) e.g. to feed spindle PWM to a VFD
- and pins for the second opamp left over
- 12V to 5 V power stage and configurable supply for the blackpill and all the circuitry
- ... with the idea to use USB-insulator for EMC stability
- 3x 8 channel 74HCT541 3.3-to-5V-buffer
- 3x 7 channel 50V/300mA(if I remember right...) open collector ULN2003A LED / Relais /whatever-driver
- 2/3 of them prerouted for 4 Axes 
- output stages with the optocoupler-input of DMT556 (and bretheren) power stepper drivers in mind
- sadly no board space left for "StepSmart" format drivers, so they might go to a breadboard and be connected by duPonts
- 1/3 digital outs prerouted for relais control with indicator LEDs
- all of the output stages, including unused lines, can be reconfigured by pin sockets

Machines in my Garage waiting to be driven:
- TTC450 (waits for recent grbl-HAL and gantry auto squaring)
- material for a steel-capable gantry mill of a bit larger size
- a halfway designed plasma cutter, with THC control
- an age old lathe, where I'd like to implement spindle sync for thread cutting
- an even older really heavy mill

My first idea of a "one size fitting all" approach was to go for a controller and board large enough to meet every need.
This board is my next learning step and expected to deliver the joy of success at shorter intervals.

**Comment by wolfgangr** [comment]
... time for some pictures

hm ... looks like I cannot add files here???
When I try the obvious - described here
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files
I see a short flash of a line, disappering right again.
Brower config? Account config?

**Comment by wolfgangr** [comment]
Please find the images in my fork of Terje's Nucleo board.

![3D footage populated](https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/blackpill_CNC_BoB_3D.jpg?raw=true)

This View is generated by KiCAD.  
The blackpill and the screw connectors (2,54 pitch) are missing since I used proprietary footage.  
All the jumper pins are shown.  
The board is assembled by JCLPCB with the SMD-components only.  
So, connectors not needed for a specific implementation might e.g. be omitted or permanently briged by soldered wires.

**Comment by wolfgangr** [comment]
This is top view:
![top view](https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/blackpill_CNC_BoB_top_view.png?raw=true)

As you can see, I'm still in the process of learning to edit silk mask ;-)
At least the outside connectors and most important jumper groups are marked.
But I'm afraid that only the combination of silk mask, schematics and my memories suffice to identify all the jumper settings required (well, at least I hope so....)

So, if anybody intents to use my design, please don't hesitate to ask.

**Comment by wolfgangr** [comment]
simplified layout:  
  
  
![simplified layout](https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/blackpill_CNC_BoB_layout.png?raw=true)

... available here as pdf for the details:
https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/blackpill_CNC_BoB_layout.pdf

**Comment by wolfgangr** [comment]
Last but not least the schematics  

![schematics preview](https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/bp-breakout-schematics-01.png?raw=true)

... here as pdf for details  
https://github.com/wolfgangr/CNC_Breakout_STM32F4xx/blob/main/design/media/bp-breakout-schematics-01.pdf

**Comment by terjeio** [comment]
Nice, can't wait to hear about how using this board for your machines works out.
And if someone else is going to use this design then the board map should be added to the STM32F4xx repo.

**Reply by wolfgangr** [reply]
Oh, forgot to say that I used blackpill_alt2_map.h as template for pin assignment  
so it is already there :-)
